instrumentation/java-classloader/src/main/java/io/opentelemetry/instrumentation/auto/javaclassloader/ResourceInjectionInstrumentation.java

```diff
@@ -0,0 +1,150 @@
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+/**
+ * Instruments {@link ClassLoader} to have calls to get resources intercepted and check our map of
+ * helper resources that is filled by instrumentation when they need helpers.
+ */
+@AutoService(Instrumenter.class)
+public class ResourceInjectionInstrumentation extends Instrumenter.Default {
+
+  public ResourceInjectionInstrumentation() {
+    super("class-loader");
+  }
+
+  @Override
+  public String[] helperClassNames() {
+    return new String[] {
+      "io.opentelemetry.instrumentation.auto.javaclassloader.BytesUtil",
+    };
+  }
+
+  @Override
+  public ElementMatcher<? super TypeDescription> typeMatcher() {
+    return extendsClass(named("java.lang.ClassLoader"));
+  }
+
+  @Override
+  public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+    Map<ElementMatcher<? super MethodDescription>, String> transformers = new HashMap<>();
+    transformers.put(
+        isMethod().and(named("getResource")).and(takesArguments(String.class)),
+        ResourceInjectionInstrumentation.class.getName() + "$GetResourceAdvice");
+
+    transformers.put(
+        isMethod().and(named("getResources")).and(takesArguments(String.class)),
+        ResourceInjectionInstrumentation.class.getName() + "$GetResourcesAdvice");
+    return transformers;
+  }
+
+  public static class GetResourceAdvice {
+    @Advice.OnMethodExit(onThrowable = Throwable.class)
+    public static void onExit(
+        @Advice.This ClassLoader classLoader,
+        @Advice.Argument(0) String name,
+        @Advice.Return(readOnly = false) URL resourceUrl) {
+      if (resourceUrl != null) {
+        // Give their classloader precedence.
+        return;
+      }
+
+      resourceUrl = HelperResources.load(classLoader, name);
+    }
+  }
+
+  public static class GetResourcesAdvice {
+    @Advice.OnMethodExit(onThrowable = Throwable.class)
+    public static void onExit(
+        @Advice.This ClassLoader classLoader,
+        @Advice.Argument(0) String name,
+        @Advice.Return(readOnly = false) Enumeration<URL> resources) {
+      URL helper = HelperResources.load(classLoader, name);
+      if (helper == null) {
+        return;
+      }
+
+      if (!resources.hasMoreElements()) {
+        resources = Collections.enumeration(Collections.singleton(helper));
+        return;
+      }
+
+      // TODO(anuraaga): Don't inline BytesUtil.toByteArray after understanding AccessError when
+      // trying to use the class from bootstrap classloader.
```

Comment:
`BytesUtil` class and its `toByteArray` method both need to be `public` b/c the advice is inlined, so it's really in a different package